src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java

```diff
@@ -210,74 +220,176 @@ private boolean checkCurrentResourcesStatus(List<LockableResource> selected,
 				needToWait = true;
-				if (context != null) {
-					r.queueAdd(context);
-					break;
-				}
+				break;
 			}
 		}
 		if (!needToWait) {
 			for (LockableResource r : resources) {
 				r.unqueue();
 				r.setBuild(build);
-				if (context != null) {
-					LockStepExecution.proceed(context, r.getName(), inversePrecedence);
-					break;
+			}
+			if (context != null) {
+				// since LockableResource contains transient variables, they cannot be correctly serialized
+				// hence we use their unique resource names
+				List<String> resourceNames = new ArrayList<String>();
+				for (LockableResource resource : resources) {
+					resourceNames.add(resource.getName());
 				}
+				LockStepExecution.proceed(resourceNames, context, logmessage, inversePrecedence);
 			}
 		}
 		save();
 		return !needToWait;
 	}
 	
+	private synchronized void freeResources(List<String> unlockResourceNames, Run<?, ?> build) {
+		for (String unlockResourceName : unlockResourceNames) {
+			for (LockableResource resource : this.resources) {
+				if (resource.getName().equals(unlockResourceName)) {
+					if (resource == null || (resource.getBuild() != null && build.getExternalizableId().equals(resource.getBuild().getExternalizableId()))) {
+						// No more contexts, unlock resource
+						resource.unqueue();
+						resource.setBuild(null);
+					}
+				}
+			}
+		}
+	}
+
 	public synchronized void unlock(List<LockableResource> resourcesToUnLock,
 			Run<?, ?> build, @Nullable StepContext context) {
-		unlock(resourcesToUnLock, build, context, false);
+		unlock(resourcesToUnLock, build, context, null, false);
 	}
 
-	public synchronized void unlock(List<LockableResource> resourcesToUnLock,
-			Run<?, ?> build, @Nullable StepContext context, boolean inversePrecedence) {
-		for (LockableResource r : resourcesToUnLock) {
-			// Search the resource in the internal list to unlock it
-			for (LockableResource internal : resources) {
-				if (internal.getName().equals(r.getName())) {
-					if (build == null ||
-							(internal.getBuild() != null && build.getExternalizableId().equals(internal.getBuild().getExternalizableId()))) {
-						// this will remove the context from the queue and setBuild(null) if there are no more contexts
-						StepContext nextContext = internal.getNextQueuedContext(inversePrecedence);
-						if (nextContext != null) {
-							try {
-								internal.setBuild(nextContext.get(Run.class));
-							} catch (Exception e) {
-								throw new IllegalStateException("Can not access the context of a running build", e);
-							}
-							LockStepExecution.proceed(nextContext, internal.getName(), inversePrecedence);
-						} else {
-							// No more contexts, unlock resource
-							internal.unqueue();
-							internal.setBuild(null);
-						}
+	public synchronized void unlock(@Nullable List<LockableResource> resourcesToUnLock,
+			Run<?, ?> build, @Nullable StepContext context,
+			@Nullable List<String> resourceNamesToUnLock, boolean inversePrecedence) {
```

Comment:
I think this new parameter (`resourceNamesToUnLock`) is going to be mutually exclusive with the original one (`resourcesToUnLock`). A new method should be created to manage this new override.